contribute-to-tidb/code-style-and-quality-guide.md

```diff
@@ -38,3 +38,20 @@ func main() {
     fmt.Println(ping.Domain())
 }
 ```
+
+## Consider using values instead of pointers
+
+Many Go programmers over-use pointers as opposed to just using values directly.
+Go programs often return pointers to structs with a wrapper and then all functions use the struct as a pointer:
+
+``` go
+type struct S {}; func NewStruct() *S
+
+func (s *S) structMethod()
+```
+
+Using pointers for the entire method set means we can no longer differentiate by the method signature when a function mutates the struct and when it simply reads data. This increases the chances of errors due to the struct ending up in an unintended state.
```

Comment:
I think readonly is just a side effect of value copy when value receivers are used, not our goal. If it is recommended to use a value receiver for readonly purposes, the consequences may introduce too many value copies and thus affect performance.

Reply:
I always have a goal in my programming to try to make it very clear when something is mutated since this is a source of bugs.

Did you read the next section? Value copies may actually improve performance.

Reply:
Yes. Performance issues depend on the specific situation. So this is a flexible trade-off, and you cannot always use value receivers for readonly purposes.

Reply:
Of course. Perhaps this needs to be re-worded to make this clearer.

Reply:
Yep. I think most of code in TiDB use the pointer receiver style. It might cause confusions to developers how we follow the current codebase or using value receiver.

An example on our codebase is admirable, but I don't insist one.

Reply:
I think programmers should be able to look at nearby code and use a style similar to the code around them. My goal is to write things that we are not doing enough of and that people should think about more. Probably these things deserve more attention, thought, and discussion before being accepted as a guide. Probably people will find it confusing if it is a guide but they don't see the pattern followed often in our code.

Reply:
> My goal is to write things that we are not doing enough of and that people should think about more.

That sounds reasonable. Let me have a think.

Reply:
@gregwebs I agree with your statement and pending to approve the PR. One more question,

> The main reason not to do this is to be able to meet the requirements of an interface.

This sounds like "the only reason" and maybe we use pointer receiver for mutation, for reduce value copying also?